rules/rbe_repo/rbe_repo_spec.bzl

```diff
@@ -0,0 +1,435 @@
+
+This file will be read when a user wants to use checked-in configs.
+It will be modified when a user wants to generate and export to the
+output_base configs to be later on used as checked-in configs.
+
+The versions.bzl file should (initially) contain exactly these definitions:
+
+def configs():
+    return []
+
+DEFAULT_TOOLCHAIN_CONFIG_SPEC = ""
+
+# Returns a dict with suppported Bazel versions mapped to the config version to use.
+def bazel_to_config_versions():
+    return {
+    }
+
+# sha256 digest of the latest version of the toolchain container.
+LATEST = ""
+
+# Map from sha256 of the toolchain container to corresponding major container
+# versions.
+def container_to_config_spec_names():
+    return {
+    }
+
+def rbe_repo_gen_spec():
+    return struct(
+        bazel_to_config_spec_names_map = bazel_to_config_versions,
+        container_to_config_spec_names_map = container_to_config_spec_names,
+        default_toolchain_config_spec = DEFAULT_TOOLCHAIN_CONFIG_SPEC,
+        latest_container = LATEST,
+        toolchain_config_specs = configs,
+    )
+
+
+This bzl file also provides utils to convert fields of the
+rbe_repo_gen_spec structs in the versions.bzl to a format that can be passed
+to the _rbe_autoconfig_impl. This is because we cannot pass
+structs to _rbe_autoconfig_impl. 
+Defs 'config_to_string_lists' and 'string_lists_to_config' in this file
+provide a way to convert an toolchain_config_specs in the rbe_repo_gen_spec
+from structs to lists and from lists to structs.
+"""
+
+# TODO(nlopezgi): move to using versions.bzl
+# once migration for toolchain config release process is complete
+load(
+    "//configs/ubuntu16_04_clang:versions_to_migrate.bzl",
+    "rbe_repo_gen_spec",
+)
+
+_SEPARATOR = ":::"
+
+REPO_SPEC_KEYS = [
+    "container_repo",
+    "container_registry",
+    "output_base",
+    "repo_name",
+    "rbe_repo_gen_spec",
+]
+
+CONFIG_SPEC_FIELDS = [
+    "bazel_to_config_spec_names_map",
+    "container_to_config_spec_names_map",
+    "default_toolchain_config_spec",
+    "latest_container",
+    "toolchain_config_specs",
+]
+
+def default_rbe_repo_spec():
+    return {
+        "repo_name": "bazel_toolchains",
+        "output_base": "configs/ubuntu16_04_clang",
+        "container_repo": "google/rbe-ubuntu16-04",
+        "container_registry": "marketplace.gcr.io",
+        "rbe_repo_gen_spec": rbe_repo_gen_spec(),
+    }
+
+def validate_rbe_repo_spec(name, rbe_repo_spec):
+    """ Validates the given rbe_repo_spec is 
+
+Should only be called from the rbe_autoconfig macro
+in //rules/rbe_repo.bzl
+
+    Args:
+      name: Name of the rbe_autoconfig rule
+      rbe_repo_spec: The rbe_repo_spec to validate
+
+    """
+    _validate_repo_spec_keys(name, rbe_repo_spec)
+    _validate_rbe_repo_gen_spec(name, rbe_repo_spec)
+
+# Validates the rbe_repo_spec is a struct and has all required top level keys
+def _validate_repo_spec_keys(name, rbe_repo_spec):
```

Comment:
Quite a bit of repetition in these functions. I wonder if this could be refactored into yet another "type schema" that maps from expected attribute names to their types and uses that to specify the expected type for the various structs being validated, marshaled & unmarshaled.

So for example a type schema could be
```
config_repo_type_schema = {
"bazel_to_config_spec_names_map": "dict",
 ...
}
```

Also, all this parsing and validation appears too heavyweight for Starlark in my opinion. Should we consider creating these configurations in other format like YAML, etc and have a utility in Go/Python etc do the high level parsing and validation and generate Starlark files with the processed & validated data. This way we avoid bending over backwards trying to work with Starlark's limitations when doing complex operations like schema validation. We generated Starlark files would then be available for import to other rules that want to use them. Doesn't have to be part of this PR but something to think about.

Reply:
managed to reduce some of the repetition. ptal.
wrt doing this in anything other than starlark, it would be really hard:
- If the data is an attr of the rule we have to process it in the analysis phase which will have to be starlark
- If we move this to outside of attrs that are passed to the repo rule, then changes to the underlying file will not trigger rebuilds as this is a repo rule and deps cannot be really expressed.